core/iwasm/common/wasm_memory.c

```diff
@@ -387,69 +494,92 @@ wasm_runtime_get_shared_heap(WASMModuleInstanceCommon *module_inst_comm)
 
 static bool
 is_app_addr_in_shared_heap(WASMModuleInstanceCommon *module_inst,
-                           bool is_memory64, uint64 app_offset, uint32 bytes)
+                           bool is_memory64, uint64 app_offset, uint32 bytes,
+                           WASMSharedHeap **target_heap)
 {
-    WASMSharedHeap *heap = get_shared_heap(module_inst);
+    WASMSharedHeap *heap = get_shared_heap(module_inst), *cur;
+    uint64 shared_heap_start, shared_heap_end;
 
     if (!heap) {
-        return false;
+        goto fail;
     }
 
     if (bytes == 0) {
         bytes = 1;
     }
 
-    if (!is_memory64) {
-        if (app_offset >= heap->start_off_mem32
-            && app_offset <= UINT32_MAX - bytes + 1) {
-            return true;
-        }
+    /* Early stop for app start address not in the shared heap(chain) at all */
+    shared_heap_start =
+        is_memory64 ? heap->start_off_mem64 : heap->start_off_mem32;
+    shared_heap_end = is_memory64 ? UINT64_MAX : UINT32_MAX;
+    if (app_offset < shared_heap_start
+        || app_offset > shared_heap_end - bytes + 1) {
+        goto fail;
     }
-    else {
-        if (app_offset >= heap->start_off_mem64
-            && app_offset <= UINT64_MAX - bytes + 1) {
-            return true;
+
+    /* Find the exact shared heap that app addr is in */
+    if (target_heap) {
+        for (cur = heap; cur; cur = cur->chain_next) {
+            shared_heap_start =
+                is_memory64 ? cur->start_off_mem64 : cur->start_off_mem32;
+            shared_heap_end = shared_heap_start - 1 + cur->size;
+            if (app_offset >= shared_heap_start
+                && app_offset <= shared_heap_end - bytes + 1) {
+                *target_heap = cur;
+                return true;
+            }
```

Comment:
- Need to handle the situation if `app_offset` is invalid
- `if (target_heap) { *target_heap = cur; }`